File: src/graphClient.js
const msal = require('@azure/msal-node');
const fetch = require('node-fetch');
const zlib = require('zlib');
const AdmZip = require('adm-zip');

const GRAPH_URL = 'https://graph.microsoft.com/v1.0';

class GraphClient {
  constructor(options) {
    if (options.accessToken) {
      // Delegated: caller provides a ready access token
      this._getToken = async () => options.accessToken;
    } else {
      // App-only: client credentials flow (used by scheduler)
      const msalApp = new msal.ConfidentialClientApplication({
        auth: {
          clientId: options.clientId,
          clientSecret: options.clientSecret,
          authority: `https://login.microsoftonline.com/${options.tenantId}`,
        },
      });
      this._getToken = async () => {
        const result = await msalApp.acquireTokenByClientCredential({
          scopes: ['https://graph.microsoft.com/.default'],
        });
        if (!result || !result.accessToken) throw new Error('Token acquisition failed');
        return result.accessToken;
      };
    }
  }

  async _headers() {
    return { Authorization: `Bearer ${await this._getToken()}` };
  }

  async getDmarcMessages(mailbox, folder = 'Inbox', limit = 100) {
    // $filter and $orderby cannot be combined on mail endpoints (InefficientFilter).
    // Filter by hasAttachments here; receivedDateTime sort is done client-side.
    const params = new URLSearchParams({
      $top: limit,
      $select: 'id,subject,receivedDateTime,from,hasAttachments',
      $filter: 'hasAttachments eq true',
    });
    const url = `${GRAPH_URL}/users/${mailbox}/mailFolders/${folder}/messages?${params}`;
    const resp = await fetch(url, { headers: await this._headers(), timeout: 30000 });
    if (!resp.ok) {
      const body = await resp.text();
      throw new Error(`Graph API ${resp.status}: ${body}`);
    }
    const data = await resp.json();
    return (data.value || [])
      .filter(m => _isDmarcSubject(m.subject || ''))
      .sort((a, b) => new Date(b.receivedDateTime) - new Date(a.receivedDateTime));
  }

  async getXmlAttachments(mailbox, messageId) {
    const url = `${GRAPH_URL}/users/${mailbox}/messages/${messageId}/attachments`;
    const resp = await fetch(url, { headers: await this._headers(), timeout: 30000 });
    if (!resp.ok) throw new Error(`Graph API ${resp.status}`);
    const data = await resp.json();

    const xmlPayloads = [];
    for (const att of data.value || []) {
      const name = (att.name || '').toLowerCase();
      const raw = Buffer.from(att.contentBytes || '', 'base64');
      if (name.endsWith('.xml')) {
        xmlPayloads.push(raw);
      } else if (name.endsWith('.gz')) {
        xmlPayloads.push(zlib.gunzipSync(raw));
      } else if (name.endsWith('.zip')) {
        const zip = new AdmZip(raw);
        for (const entry of zip.getEntries()) {
          if (entry.entryName.toLowerCase().endsWith('.xml')) {
            xmlPayloads.push(entry.getData());
          }
        }
      }
    }
    return xmlPayloads;
  }
}

function _isDmarcSubject(subject) {
  const s = subject.toLowerCase();
  return ['dmarc', 'report domain:', 'aggregate report'].some(kw => s.includes(kw));
}

module.exports = { GraphClient };

File: dmarc_dashboard/graph_client.py
import base64
import gzip
import io
import zipfile
from typing import List, Optional

import msal
import requests


class GraphClient:
    GRAPH_URL = "https://graph.microsoft.com/v1.0"

    def __init__(self, tenant_id: str, client_id: str, client_secret: str):
        self._app = msal.ConfidentialClientApplication(
            client_id,
            authority=f"https://login.microsoftonline.com/{tenant_id}",
            client_credential=client_secret,
        )

    def _token(self) -> str:
        result = self._app.acquire_token_silent(
            ["https://graph.microsoft.com/.default"], account=None
        )
        if not result:
            result = self._app.acquire_token_for_client(
                scopes=["https://graph.microsoft.com/.default"]
            )
        if "access_token" not in result:
            raise RuntimeError(
                f"Token acquisition failed: {result.get('error_description', result.get('error'))}"
            )
        return result["access_token"]

    def _headers(self) -> dict:
        return {"Authorization": f"Bearer {self._token()}"}

    def get_dmarc_messages(
        self, mailbox: str, folder: str = "Inbox", limit: int = 100
    ) -> List[dict]:
        """Return emails that are likely DMARC aggregate reports."""
        url = f"{self.GRAPH_URL}/users/{mailbox}/mailFolders/{folder}/messages"
        params = {
            "$top": limit,
            "$select": "id,subject,receivedDateTime,from,hasAttachments",
            "$filter": "hasAttachments eq true",
            "$orderby": "receivedDateTime desc",
        }
        resp = requests.get(url, headers=self._headers(), params=params, timeout=30)
        resp.raise_for_status()

        messages = resp.json().get("value", [])
        return [
            m
            for m in messages
            if _is_dmarc_subject(m.get("subject", ""))
        ]

    def get_xml_attachments(self, mailbox: str, message_id: str) -> List[bytes]:
        """Download and decompress all XML DMARC report attachments from a message."""
        url = f"{self.GRAPH_URL}/users/{mailbox}/messages/{message_id}/attachments"
        resp = requests.get(url, headers=self._headers(), timeout=30)
        resp.raise_for_status()

        xml_payloads: List[bytes] = []
        for att in resp.json().get("value", []):
            name: str = att.get("name", "").lower()
            raw = base64.b64decode(att.get("contentBytes", ""))
            if name.endswith(".xml"):
                xml_payloads.append(raw)
            elif name.endswith(".gz"):
                xml_payloads.append(gzip.decompress(raw))
            elif name.endswith(".zip"):
                with zipfile.ZipFile(io.BytesIO(raw)) as zf:
                    for zname in zf.namelist():
                        if zname.lower().endswith(".xml"):
                            xml_payloads.append(zf.read(zname))
        return xml_payloads


def _is_dmarc_subject(subject: str) -> bool:
    subject_lower = subject.lower()
    keywords = ["dmarc", "report domain:", "aggregate report"]
    return any(kw in subject_lower for kw in keywords)
